### DISCUSSION TITLE
apt-offline does not download some specific upgrade packages

### DISCUSSION THREAD

**Discussion by narayanan-cr** [rickysarraf/apt-offline#153] [Q&A] [answered]
I started with two VirtualMachines with Ubuntu 18.04 server OS - one being a clone of another. 
Here, I want to test the offline upgrades, so the intention is to keep one of the machine offline and use apt-offline tool to manage the updates and upgrades. 
I followed the steps and was able to complete the upgrades on the offline machine - But found a strange behavior, that particular packages were listed as upgradable. 
So I again repeated the process of generating "set.uris" for '--update' on offline machine, then did a 'get' and then 'install'. Next I generated 'set.uris' for '--upgrade' - While doing a 'get', apt-offline did not download any packages when I used 
"apt-offline get upgrade.sig --bundle upgrade.zip" command.
To cross verify, I ran the following commands on the machine that was online:
1. sudo  apt update
2. sudo apt-get upgrade
Now on viewing the upgradable packages, I see the same set of packages as seen on offline machine:

![image](https://user-images.githubusercontent.com/48640610/130401387-39f8f54a-3b6b-4792-b185-91a2adfa8f19.png)

So I try to run 'sudo apt-get upgrade' multiple times on the online machine and it did not do any changes.
So next I tried, "sudo apt upgrade" and the packages were downloaded and installed.. !

1. I did not understand, why the 'sudo apt upgrade' resolved the upgrades and not the sudo apt-get upgrade ?
2. Why the 'apt-offline get' did not download these packages ? (Note-I also tried dist-upgrade and still packages were not downloaded)
3. How to install these packages on the offline machine ?

**Comment by rickysarraf** [comment, marked as answer]
For point no 1, there are subtle differences in the behavior of `apt upgrade` vs `apt-get upgrade`.

For point no 2, THat's a bit of a surprise. `dist-upgrade` should work. Can you share the output of the `.sig` file ? Also, when investigating such cases, you can use the `--verbose` option. It'll give a lot more information.

For point no 3, everything looks fine to me. These are all installed packages and candidate for upgrades. Once we look at the verbose logs from point no 2, it should help ascertain the cause.

Note: Since you mentioned Ubuntu @ 18.04, it may also possibly be shipping an older version of `apt-offline`. You may want to try with the latest from https://packages.debian.org/search?keywords=apt-offline
Or if you are comfortable with, you can try the `master` branch.

**Reply by narayanan-cr** [reply]
Thanks for the explanation. The issue got resolved after using the "dist-upgrade" flag. 
By mistake, I had used a wrong *.sig file the first time and hence the thought that it did not work even with the 'dist-upgrade' flag. 
Thankyou again.
Best Regards,

**Reply by rickysarraf** [reply]
Great. I'm glad you were able to figure it out. Thanks for confirming.

**Reply by semaka** [reply]
very interesting subject, directly to the problem. I have the same problem and even with last solution I still can't collect the whole upgrade packages. How to use dist-upgrade with apt-offline to collect everything? I'm using ubuntu 20.04
This one:
`sudo apt-offline set --update --upgrade-type dist-upgrade apt-offline.sig`

does not get most of packages. How could be possible to do this?

**Reply by rickysarraf** [reply]
`--upgrade-type` is efffective only when used with `--upgrade` option. Please refer to the howto/manpage

**Reply by semaka** [reply]
I already read the howto/manage but still can't figure out to results. Do you have any Idea to try? So the question remains the same, how to use apt-offline to get the all packages ready for upgrade offline, same as the command "apt upgrade" is doing online?

**Reply by rickysarraf** [reply]
Please keep retrying until you figure out. As you can see, other users have also confirmed that it works fine on similar Ubuntu systems. There's the [howto](https://github.com/rickysarraf/apt-offline/wiki) along with a link to a video tutorial. It should help you figure out what where you are doing wrong.

If you think the documentation isn't obvious enough, patches to the documentation are welcome.

**Comment by rickysarraf** [comment]
Please see and continue the conversation in this thread. It helps
general public, current and future prospects.

https://github.com/rickysarraf/apt-offline/discussions/153


On Tue, 2022-03-29 at 08:13 +0000, Gerard T. wrote:
> Hi,
> 
>     I really face a big problem in trying to get upgrade packages
> collection for ubuntu 20.04. 
>    If I use "apt-get upgrade" I still have packages not upgraded, but
> if I use "apt upgrade" everything works nice. 
>    So if I use "apt-offline set apt-offline.sig" this command give me
> reference for getting some packages, exactly as "apt-get upgrade" do.
>    What is the command line for apt-offline to get the whole packages
> exactly as the command "apt upgrade" do???
> 
>     I already tried also this one below but without success:
> "apt-offline set --update --upgrade-type dist-upgrade apt-
> offline.sig"
> 
>     Thanks a lot,
>     Best regards,
>     Gerard
> 

-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
Necessity is the mother of invention
आव्श्यक्ता आविष्कार की जननी है

**Comment by semaka** [comment]
I really want to retry as many times it's necessary but I need some clues from people who did it. Because there is just a confirmation for using dist-upgrade to solve the problem, but not indicated how to use it. 
This is why I gave the example how I understood to use it, but not working. I will still try, but if somebody see my request which put this in practice, maybe will help me, because I can't find the solution over internet.
BR,

    On Tuesday, 29 March 2022, 12:38:21 EEST, Ritesh Raj Sarraf ***@***.***> wrote:  
 
 


Please keep retrying until you figure out. As you can see, other users have also confirmed that it works fine on similar Ubuntu systems. There's the howto along with a link to a video tutorial. It should help you figure out what where you are doing wrong.

If you think the documentation isn't obvious enough, patches to the documentation are welcome.

**Comment by semaka** [comment]
Thank you for suggestion, I'm already there, semaka is me. By the way, your suggested link is for clasical apt-offline use which substitute "apt-get upgrade" but not for special apt-offline which substitute "apt upgrade". Between those two it is a big difference but can't find a practical solution.
Thanks for your trying to help in this,

    On Tuesday, 29 March 2022, 14:33:15 EEST, Ritesh Raj Sarraf ***@***.***> wrote:  
 
 

Please see and continue the conversation in this thread. It helps
general public, current and future prospects.

https://github.com/rickysarraf/apt-offline/discussions/153


On Tue, 2022-03-29 at 08:13 +0000, Gerard T. wrote:
> Hi,
> 
>     I really face a big problem in trying to get upgrade packages
> collection for ubuntu 20.04. 
>    If I use "apt-get upgrade" I still have packages not upgraded, but
> if I use "apt upgrade" everything works nice. 
>    So if I use "apt-offline set apt-offline.sig" this command give me
> reference for getting some packages, exactly as "apt-get upgrade" do.
>    What is the command line for apt-offline to get the whole packages
> exactly as the command "apt upgrade" do???
> 
>     I already tried also this one below but without success:
> "apt-offline set --update --upgrade-type dist-upgrade apt-
> offline.sig"
> 
>     Thanks a lot,
>     Best regards,
>     Gerard
> 

-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
Necessity is the mother of invention
आव्श्यक्ता आविष्कार की जननी है

**Reply by semaka** [reply]
So...is it any explanation why apt-offline get just part of packages like "apt-get upgrade" do, and not getting all available packages like "apt upgrade" do? Does anybody know how apt-offline works with apt to see this problem? Here in discussion to solve the problem we don't have a solution as line command just a confirmation which can't be confirmed by me.
I have no solution yet, how to upgrade in offline mode in such way that if I connected to internet after to see that no package is available for upgrade.

**Reply by rickysarraf** [reply]
Under `apt-offline set` you have this option `--apt-backend`. It is well documented in the help section of the command.

```
rrs@priyasi:~/.../apt-offline (master)$ ./apt-offline set -h
usage: apt-offline set [-h] [--verbose] [--quiet] [--simulate] [--install-packages [PKG ...]] [--install-src-packages [SOURCE PKG ...]] [--src-build-dep] [--release release_name] [--update]
                       [--upgrade] [--upgrade-type upgrade] [--generate-changelog] [--apt-backend apt-get]
                       apt-offline.sig

positional arguments:
  apt-offline.sig       Generate a signature file

optional arguments:
  -h, --help            show this help message and exit
  --verbose             Enable verbose messages
  --quiet               Enable quiet mode
  --simulate            Just simulate. Very helpful when debugging
  --install-packages [PKG ...]
                        Packages that need to be installed
  --install-src-packages [SOURCE PKG ...]
                        Source Packages that need to be installed
  --src-build-dep       Install Build Dependency packages for requested source packages
  --release release_name
                        Release target to install packages from
  --update              Generate Signature to update APT Database
  --upgrade             Generate Signature of packages to be upgraded
  --upgrade-type upgrade
                        Type of upgrade to do. Use one of upgrade, dist-upgrade, dselect-ugprade
  --generate-changelog  Generate changelog of the version to be downloaded
  --apt-backend apt-get
                        APT backend to use. One of: apt, apt-get, python-apt
```

**Reply by semaka** [reply]
wow...this could be the solution, let me test and I'll return with results. Thanks a lot.

**Reply by semaka** [reply]
YES, this was the clue. I tested and works great. The magic command was:
```
apt-offline set apt-backend apt bundle.sig
apt-offline get --bundle bundle.zip bundle.sig
```

The rest is simple. I did the test and now I can say that apt-offline is magic good.
Thank you,

**Reply by rickysarraf** [reply]
If you think the documentation could be improved, PRs welcome.

**Comment by luisprox** [comment]
I didn't try the `apt-backend` option, but I discovered I had to run it twice. There is also  a problem with v1.8.4 which is installed from `apt install apt-offline` in Ubuntu 22. So I manually downloaded the 1.8.5 deb package from pkg.org. [Here](https://askubuntu.com/a/1503651/1073024) I explain my procedure.
